Receive Headers Limits RSKIP (#200)
rsksmart/RSKIPs: IPs/RSKIP200.md

# Receive headers limits

|RSKIP          |200           |
| :------------ |:-------------|
|**Title**      |Receive headers limits |
|**Created**    |08-JAN-21 |
|**Author**     |PGP & MI |
|**Purpose**    |Sec |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

This RSKIP proposes to make the current `receiveHeaders` bridge method private while creating a new public method `receiveHeader` that allows registering only one header at a time.

## Motivation

Bridge precompiled contract includes a function `receiveHeaders` that receives an array of headers to be stored. This function could be exposed to attacks.
Since powpeg nodes still need to send a group of headers to the Bridge, the current `receiveHeaders` function is made private. A new function `receiveHeader` is created, which only allows registering one header at a time, to be used by external users.

## Specification

To implement this change we suggest making the following modifications:

1. Create a new method in the bridge: 
    ```

>     receive_header(bytes header) returns (int256 returnValue)

Reviewer: ```suggestion
    receiveHeader(bytes header) returns (int256 returnValue)
```

    ```

> 2. Make current `receive_headers` function private

Reviewer: ```suggestion
2. Make current `receiveHeaders` function private
```

> ### `receive_header` validations

Reviewer: ```suggestion
### `receiveHeader` validations
```

1) Header has expected size. 
2) Header is not previously saved in storage.

> 3) Time between calls to receive_header is larger than the minimum time between calls allowed. This value is configured for each network.

Reviewer: ```suggestion
3) Time between calls to receiveHeader is larger than the minimum time between calls allowed. This value is configured for each network.
```

4) Header previous block must be found in storage.
5) Block depth of the received header is not above the maximum depth allowed. This value is configured for each network.

> Every time that a block header is accepted, the timestamp is saved in storage to validate that no more calls are made to `receive_header` until the minimum configured time has elapsed.

Reviewer: ```suggestion
Every time that a block header is accepted, the timestamp is saved in storage to validate that no more calls are made to `receiveHeader` until the minimum configured time has elapsed.
```

When the header is successfully saved, the method returns the value `0`. 
 

> ### `receive_header` return values

Reviewer: ```suggestion
### `receiveHeader` return values
```

-  0  OK
- -1  RECEIVE_HEADER_CALLED_TOO_SOON
- -2  RECEIVE_HEADER_BLOCK_TOO_OLD
- -3  RECEIVE_HEADER_CANT_FOUND_PREVIOUS_BLOCK
- -4  RECEIVE_HEADER_BLOCK_PREVIOUSLY_SAVED
- -20 RECEIVE_HEADER_ERROR_SIZE_MISTMATCH
- -99 RECEIVE_HEADER_UNEXPECTED_EXCEPTION

> ### `receive_headers` changes

Reviewer: ```suggestion
### `receiveHeaders` changes
```

> After the activation of this RSKIP, `receive_headers` can be called only by powpeg nodes.

Reviewer: ```suggestion
After the activation of this RSKIP, `receiveHeaders` can be called only by members of the Federation.
```

> # Bridge performance improvement

Reviewer 2: ```suggestion
# Receive headers limits
```

|RSKIP          |200           |
| :------------ |:-------------|
|**Title**      |Receive headers limits |
|**Created**    |08-JAN-21 |

> |**Author**     |PGP |

Reviewer 2: ```suggestion
|**Author**     |PGP & MI |
```

|**Purpose**    |Sca,USa,Sec |
|**Layer**      |Core |
|**Complexity** |2 |
|**Status**     |Draft |

## Abstract

> This RSKIP proposes to include new functionality in the bridge precompiled contract to receive one header every time that this function is called and save this header in the internal storage of the bridge. 
> This also included a modification in the actual method to receive headers in the same precompiled contract. These methods receive an array of headers to be stored. The modification for this method implies not allow to be called for external users after this RSKIP. 

Reviewer 2: ```suggestion
This RSKIP proposes to make the current `receiveHeaders` bridge method private while creating a new public method `receiveHeader` that allows registering only one header at a time.
```

## Motivation

Bridge precompiled contract includes a function to can receive an array of heads to be stored. This function could be exposed to attacks. To avoid this, some new limits are included. 

> Powpeg nodes need to send a group of headers anyway. 
> Since one of these new limits is to receive only one header per time, and these limits are needed only for external users, a new method with the new limits is created and the original one was modified to be used only by powpeg nodes. 

Reviewer 2: ```suggestion
Since powpeg nodes still need to send a group of headers to the Bridge, the current `receiveHeaders` function is made private. A new function `receiveHeader` is created, which only allows registering one header at a time, to be used by external users.
```

## Specification

To implement this change we suggest making the following modifications:

New function:
```
 reveive_header(bytes header) returns (int256 returnValue)
```

> Modification in the actual receive_headers function:
> After this RSKIP the function needs to return to private. 

Reviewer 2: ```suggestion
Make current `receive_headers` function private
```

> ### reveive_header limits

Reviewer 2: ```suggestion
### receive_header validations
```

> 1) The size of the parameter received is verified. These limits also exist in receive_headers function. 

Reviewer 2: What size should it have?

2) Header is not previously saved in storage.

> 3) Time between calls to receive_headers is too short. The time specified between calls to this function is got using the constants of each network. The constant used to compare is named minSecondsBetweenCallsToReceiveHeader.

Reviewer 2: ```suggestion
3) Time between calls to receive_headers is too short.
```
I don't think we should include specific implementation details in RSKIPs

> 3) Time between calls to receive_headers is too short. The time specified between calls to this function is got using the constants of each network. The constant used to compare is named minSecondsBetweenCallsToReceiveHeader.

Reviewer 2: The fact there is a time limit between calls should be included in the Specification maybe

4) The previous block can not be found in storage. Block header received can not be connected with the previous one. 
5) The depth of these block headers is too old. The height of the block is compared with the maxDepthBlockchainAccepted. This value is also saved as a contact for each network. 

Every time that a block header is accepted, the timestamp is saved in BridgeStoragedProvider to can be compared the next time that receive_header is call. 

When the header was saved, the method returns 0. 
 
### receive_header return values
-  0  OK
- -1  RECEIVE_HEADER_CALLED_TOO_SOON
- -2  RECEIVE_HEADER_BLOCK_TOO_OLD
- -3  RECEIVE_HEADER_CANT_FOUND_PREVIOUS_BLOCK
- -4  RECEIVE_HEADER_BLOCK_PREVIOUSLY_SAVED
- -20 RECEIVE_HEADER_ERROR_SIZE_MISTMATCH
- -99 RECEIVE_HEADER_UNEXPECTED_EXCEPTION

> ### reveive_headers changes

Reviewer 2: ```suggestion
### receive_headers changes
```

> After the activation of this RSKIP, the function can be called only by powpeg node. This modification is needed in the function that calculates if the method is public or not. This function returns true after RSKIP124, and must return false after this RSKIP.

Reviewer 2: ```suggestion
After the activation of this RSKIP, `receive_headers` can be called only by powpeg node.
```
